File: 783.cpp
// ref: https://leetcode.com/problems/minimum-distance-between-bst-nodes/discuss/114834/C%2B%2BJavaPython-Inorder-Traversal-O(N)-time-Recursion

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pre{-1}, res{INT_MAX};
    int minDiffInBST(TreeNode* root) {
        if (root->left) {
            minDiffInBST(root->left);
        }
        if (pre >= 0) {
            res = min(res, root->val - pre);
        }
        pre = root->val;
        if (root->right) {
            minDiffInBST(root->right);
        }
        return res;
    }
};

File: 783.py
# ref: https://leetcode.com/problems/minimum-distance-between-bst-nodes/discuss/114834/C%2B%2BJavaPython-Inorder-Traversal-O(N)-time-Recursion

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        vals = [float('-inf'), float('inf')] # pre and res
        def find(node):
            if node.left:
                find(node.left)
            vals[1] = min(vals[1], node.val - vals[0])
            vals[0] = node.val
            if node.right:
                find(node.right)
        find(root)
        return vals[1]
